utils/core/src/serde/byte_reader.rs

```diff
@@ -0,0 +1,166 @@
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if an array of the specified length could not be read
+    /// from `self`.
+    fn read_array<const N: usize>(&mut self) -> Result<[u8; N], DeserializationError>;
+
+    /// Checks if it is possible to read at least `num_bytes` bytes from this ByteReader
+    ///
+    /// # Errors
+    /// Returns an error if, when reading the requested number of bytes, we go beyond the
+    /// boundaries of the array
+    fn check_eor(&self, num_bytes: usize) -> Result<(), DeserializationError>;
+
+    /// Returns true if there are more bytes left to be read from `self`.
+    fn has_more_bytes(&self) -> bool;
+
+    // PROVIDED METHODS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a boolean value read from `self` consuming 1 byte from the reader.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u16 value could not be read from `self`.
+    fn read_bool(&mut self) -> Result<bool, DeserializationError> {
+        let byte = self.read_u8()?;
+        match byte {
+            0 => Ok(false),
+            1 => Ok(true),
+            _ => Err(DeserializationError::InvalidValue(format!(
+                "{byte} is not a boolean value"
+            ))),
+        }
+    }
+
+    /// Returns a u16 value read from `self` in little-endian byte order.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u16 value could not be read from `self`.
+    fn read_u16(&mut self) -> Result<u16, DeserializationError> {
+        let bytes = self.read_array::<2>()?;
+        Ok(u16::from_le_bytes(bytes))
+    }
+
+    /// Returns a u32 value read from `self` in little-endian byte order.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u32 value could not be read from `self`.
+    fn read_u32(&mut self) -> Result<u32, DeserializationError> {
+        let bytes = self.read_array::<4>()?;
+        Ok(u32::from_le_bytes(bytes))
+    }
+
+    /// Returns a u64 value read from `self` in little-endian byte order.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u64 value could not be read from `self`.
+    fn read_u64(&mut self) -> Result<u64, DeserializationError> {
+        let bytes = self.read_array::<8>()?;
+        Ok(u64::from_le_bytes(bytes))
+    }
+
+    /// Returns a u128 value read from `self` in little-endian byte order.
+    ///
+    /// # Errors
+    /// Returns a [DeserializationError] if a u128 value could not be read from `self`.
+    fn read_u128(&mut self) -> Result<u128, DeserializationError> {
+        let bytes = self.read_array::<16>()?;
+        Ok(u128::from_le_bytes(bytes))
+    }
+
+    /// Reads a deserializable value from `self`.
+    ///
+    /// # Panics
+    /// Panics if the value could not be read from `self`.
+    fn read<D: Deserializable>(&mut self) -> Result<D, DeserializationError>
```

Comment:
optional nit: it is better to keep a single style to declare generics. Either we put all the trait requirements inside the `where`, or we declare them with the generics.

Here we have both. The community currently favors `where` as a more readable way to quickly check the trait requirements, so I'd recommend this.

Reply:
Agreed! Fixed!